### DISCUSSION TITLE
Correct usage patterns

### DISCUSSION THREAD

**Discussion by lewisxy** [capnproto/capnproto#1849] [Q&A] [answered]
I have a few questions while using capnproto RPC.

1. When implementing `Server`, I would like to divide the work into several pieces by passing `CallContext` around. Should I pass it by value or reference? In other words, is copy allowed and is it expensive to copy? What about `Reader`, `Builder`, `Request`, `Response` object?
2. How should I correctly throw a runtime exception from the server so that the client can see it? I could design my protocol to explicitly include an `error` text field in return value, but I am wondering if there are better approach.
```c++
kj::Promise<void> MyInterfaceImpl::doSomething(DoSomethingContext context) {
  doFirstThing(context);
  // 1. void doFirstThing(DoSomethingContext& ctx); or void doFirstThing(DoSomethingContext ctx); ?
  if (!doLastThing(context)) {
    throw "doSomething failed with blah blah blah reason";
    // 2. this feels wrong, the client did NOT received the error message
  }
  return kj::READY_NOW;
}
```
3. Suppose I have the following interface and implementation. I am assuming `capnp::Data::Reader` does not copy (or take ownership of) the data (represented by `buf`). Does `setData` copy the data (from `buf` vector)? If not, what's the lifetime requirement for `buf`? The below example is contrived. In practice, `buf` may or may not be a local variable or something owned by `MyInterfaceImpl`.
```
interface MyInterface {
  doSomething @0 () -> (data :Data);
}
```
```c++
kj::Promise<void> MyInterfaceImpl::doSomething(DoSomethingContext context) {
  auto builder = context.getResults();
  std::vector<uint8_t> buf;
  // populate the data somehow
  builder.setData(capnp::Data::Reader(buf.data(), buf.size()));
  return kj::READY_NOW;
}
```
4. I am working on an adaptor between our internal async types and `kj::Promise`. A simplified version is provided as the following. `Internal::Result<T>` is a wrapper contains an eventual result of type `T`. The callback function in `setCallback()` is called from another thread (managed by our internal async system) when result is ready. `get()` will return a reference of the result if exist, `ok()` will test if it's normal (i.e., no error/exception).
```c++
template <typename return_t>
kj::Promise<return_t> make_promise(typename std::shared_ptr<Internal::Result<return_t>> result) {
  // make a promise that resolve to the eventual result of T when Internal::Result<T> is ready
  auto paf = kj::newPromiseAndCrossThreadFulfiller<void>();
  result->setCallback([f = kj::mv(paf.fulfiller)] () { f->fulfill(); });
  return paf.promise.then([result]() {
    if (!result->ok()) {
      throw "not ok";
    }
    return std::move(result->get());
  });
}
```
Some example usage of `make_promise`. Assuming `Internal::DoSomething` is some internal API calls that returns a `std::shared_ptr<Internal::Result<result_t_or_whatever>>`.
```c++
kj::Promise<void> MyInterfaceImpl::doSomething(DoSomethingContext context) {
  auto result_ptr = Internal::DoSomething(/* use params in context */);
  return make_promise(result_ptr).then([context&] (result_t_or_whatever result) { /* put result in context */ });
}
```
Are there any obvious red flags in this implementation? The only uncomfortable part is the exception? It feels wrong to use `throw` inside a promise. But I am not sure how to fix it.

Thanks

**Comment by kentonv** [comment, marked as answer]
1. `CallContext` behaves like a pointer. You may pass it by value. The underlying object backing it is destroyed when the RPC call completes (or is canceled), at which point any existing `CallContext`s pointing at it become dangling pointers and should not be used again. `Reader` and `Builder` similarly are pointer-like, becoming invalid when the `MessageReader` or `MessageBuilder` they point into goes away. `Request` and `Response` own their content; they can only be passed by move, not by copy.

2. KJ-style exceptions thrown by the server will propagate to the client. KJ exceptions are meant to be used to express unrecoverable problems, or logistical problems orthogonal to the API semantics; they are NOT intended to be used as part of your API semantics. See [various](https://github.com/capnproto/capnproto/blob/v2/kjdoc/tour.md#debugging-and-observability) [documentation](https://github.com/capnproto/capnproto/blob/v2/style-guide.md#exceptions) on how to use them. If you are looking for something like "typed exceptions" or "checked exceptions" that form part of your API, you should arrange to return an error description instead of using exceptions.

3. `setFoo()` always makes a deep copy. If you actually want to reference data external to the message buffer, you need to use advanced APIs -- look at `capnp::Orphanage`.

4. You should replace your `throw` with KJ assertion macros that throw KJ exceptions, like `KJ_ASSERT(result->ok());`. Otherwise, your code looks fine. Of course, you need to create a `kj::EventLoop` (possibly via `kj::setupAsyncIo()`) and run it in the thread where you want to use `kj::Promise`-related APIs.

**Reply by lewisxy** [reply]
Thank you for the suggestions!
> The underlying object backing it is destroyed when the RPC call completes (or is canceled)

When does RPC completes or (is canceled)? From server's perspective, does it happen when I return a `kj::Promise<void>` or when that promise eventually resolved to `void` (or get rejected)?

**Reply by kentonv** [reply]
It completes when the promise resolves.

Cancellation is driven by the client. The client can request cancellation of a call at any time. If the server has opted into allowing cancellation (via the `$allowCancellation` annotation applied to the schema), then when the client requests it, the Promise will be cancelled server-side. [Here's documentation on how cancellation works with KJ promises.](https://github.com/capnproto/capnproto/blob/v2/kjdoc/tour.md#cancellation)
